src/autoscaling_alloc_buffer.jl

```diff
@@ -0,0 +1,126 @@
+const DEFAULT_INITIAL_BUFFER_SIZE = 2^27 # 128 MB
+const DEFAULT_HISTORY_SIZE = 5
+
+mutable struct AutoscalingAllocBuffer
+    main_buffer::AllocBuffer{Vector{UInt8}}
+    const additional_buffers::Vector{AllocBuffer{Vector{UInt8}}}
+    new_buffer_size::Int
+    const total_used_history::Vector{Int}
+    const max_history_size::Int
+end
+
+"""
+    AutoscalingAllocBuffer(initial_buffer_size::Int=$DEFAULT_INITIAL_BUFFER_SIZE;
+                           max_history_size=$DEFAULT_HISTORY_SIZE)
+
+Construct a `AutoscalingAllocBuffer`. This constructs an `AllocBuffer` of size `initial_buffer_size`. If an allocation (of some size `allocation_size`) is requested and there is not enough space, it will allocate a new buffer (whose size is determined by internal heuristics).
+
+When `reset_buffer!` is called, if there are additional buffers, a new larger main buffer will be created so that if similar sized allocations occur on the next run no additional buffers will be needed. It will remember the amount of allocations used in the past `max_history_size` runs to inform the size of the new main buffer, to ensure there is at least enough room for each of those past runs.
+
+This means:
+
+- `AutoscalingAllocBuffer` does not run out of memory (unlike `AllocBuffer`), since new memory will allocated on-demand when necessary
+- for repeated runs of the same size, a single contiguous buffer will be used, which should approximately match the performance of a tuned `AllocBuffer`
+- `AutoscalingAllocBuffer` can reuse allocated memory between runs like `AllocBuffer`, but without potentially OOMing like `AutoscalingAllocBuffer`. Additionally, `AutoscalingAllocBuffer` separately tracks the memory used by the main buffer vs the additional buffers allocated dynamically, so small unexpected additional allocations don't double the memory consumption (unlike a second slab being allocated).
```

Comment:
I think the ?? here is supposed to be something other than AutoscalingAllocBuffer?
```suggestion
- `AutoscalingAllocBuffer` can reuse allocated memory between runs like `AllocBuffer`, but without potentially OOMing like `??`. Additionally, `AutoscalingAllocBuffer` separately tracks the memory used by the main buffer vs the additional buffers allocated dynamically, so small unexpected additional allocations don't double the memory consumption (unlike a second slab being allocated).
```